runtime/lua/vim/_meta.lua

```diff
@@ -0,0 +1,629 @@
+      if bufnr == nil and type(k) == "number" then
+        return new_buf_opt_accessor(k)
+      end
+
+      return a.nvim_buf_get_option(bufnr or 0, k)
+    end
+
+    local function set(k, v)
+      return a.nvim_buf_set_option(bufnr or 0, k, v)
+    end
+
+    return make_meta_accessor(get, set, nil, function(k)
+      if type(k) == 'string' then
+        if win_options[k] then
+          error(string.format([['%s' is a window option, not a buffer option. See ":help %s"]], k, k))
+        elseif glb_options[k] then
+          error(string.format([['%s' is a global option, not a buffer option. See ":help %s"]], k, k))
+        end
+      end
+
+      return true
+    end)
+  end
+
+  vim.bo = new_buf_opt_accessor(nil)
+end
+
+do -- window option accessor
+  local function new_win_opt_accessor(winnr)
+    local function get(k)
+      if winnr == nil and type(k) == "number" then
+        return new_win_opt_accessor(k)
+      end
+      return a.nvim_win_get_option(winnr or 0, k)
+    end
+
+    local function set(k, v)
+      return a.nvim_win_set_option(winnr or 0, k, v)
+    end
+
+    return make_meta_accessor(get, set, nil, function(k)
+      if type(k) == 'string' then
+        if buf_options[k] then
+          error(string.format([['%s' is a buffer option, not a window option. See ":help %s"]], k, k))
+        elseif glb_options[k] then
+          error(string.format([['%s' is a global option, not a window option. See ":help %s"]], k, k))
+        end
+      end
+
+      return true
+    end)
+  end
+
+  vim.wo = new_win_opt_accessor(nil)
+end
+
+--[[
+Local window setter
+
+buffer options: does not get copied when split
+  nvim_set_option(buf_opt, value) -> sets the default for NEW buffers
+    this sets the hidden global default for buffer options
+
+  nvim_buf_set_option(...) -> sets the local value for the buffer
+
+  set opt=value, does BOTH global default AND buffer local value
+  setlocal opt=value, does ONLY buffer local value
+
+window options: gets copied
+  does not need to call nvim_set_option because nobody knows what the heck this does⸮
+  We call it anyway for more readable code.
+
+
+    Command                 global value       local value
+      :set option=value         set             set
+ :setlocal option=value          -              set
+:setglobal option=value         set              -
+--]]
+local function set_scoped_option(k, v, set_type)
+  local info = options_info[k]
+
+  -- Don't let people do setlocal with global options.
+  --    That is a feature that doesn't make sense.
+  if set_type == SET_TYPES.LOCAL and is_global_option(info) then
+    error(string.format("Unable to setlocal option: '%s', which is a global option.", k))
+  end
+
+  -- Only `setlocal` skips setting the default/global value
+  --    This will more-or-less noop for window options, but that's OK
+  if set_type ~= SET_TYPES.LOCAL then
+    a.nvim_set_option(k, v)
+  end
+
+  if set_type ~= SET_TYPES.GLOBAL then
```

Comment:
if the option is global-local then `vim.o` should not set the local value.

Reply:
I don't understand -- I thought `vim.o` was like `set`, and in the documentation it says

```
    Command                 global value       local value
      :set option=value         set             set
 :setlocal option=value          -              set
:setglobal option=value         set              -
```

Reply:
(discussed on element)